File: wozmon/wozmon.py
mem = [0] * 64 * 1024
escape = True
def echo(s): print(s, end='')

while True:
    if escape: echo('\\'); escape = False # escape
    print() # getline
    inp = input() # nextchar
    y = 0
    mode = 'xam' # setmode
    while y < len(inp):
        c = inp[y] # nextitem
        if c < '.': y += 1 # blskip-ish
        elif c == '.':
            mode = 'block'; y += 1
        elif c == ':':
            mode = 'store'; y += 1
        elif c == 'r': break # run, not implemented
        else:
            hx = 0
            ysav = y
            while y < len(inp):
                c = inp[y] # nexthex
                if c.isdigit() or 'a' <= c <= 'f':
                    hx = hx * 16 + int(c, base=16) # dig
                else:
                    break
                y += 1
            if y == ysav: # nothex
                escape = True
                break
            if mode == 'store':
                mem[st] = hx
                st += 1
                continue
            if mode == 'xam': # notstor
                st = hx # setadr
                xam = hx
            addr = True
            while True:
                if mode == 'xam': # hack
                    if addr: # nxtprnt
                        echo(f'\n{xam:04x}:')
                    echo(f' {mem[xam]:02x}') # prdata
                mode = 'xam' # xamnext
                if hx == xam:
                    break
                xam += 1
                addr = xam % 8 == 0 # mod8chk

File: wozmon/wozmon.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <ctype.h>

#define MEM_SIZE (64 * 1024)

int main() { // with help from ChatGPT 3.5
    unsigned char mem[MEM_SIZE] = {0};
    bool escape = true;
    char in[128];
    int st = 0;
    int xam = 0;

    while (true) {
        if (escape) {
            printf("\\"); // escape
            escape = false;
        }
        printf("\n"); // getline
        fgets(in, sizeof(in), stdin); // nextchar
        int y = 0;
        char mode = 'x'; // setmode, x=xam, b=block, s=store
        while (in[y] != '\0') {
            char c = in[y]; // nextitem
            if (c < '.') y++; // blskip-ish
            else if (c == '.') {
                mode = 'b';
                y++;
            } else if (c == ':') {
                mode = 's';
                y++;
            } else if (c == 'r') break; // run, not implemented
            else {
                int hx = 0;
                int ysav = y;
                while (in[y] != '\0') {
                    char c = in[y]; // nexthex
                    if (isdigit(c) || (c >= 'a' && c <= 'f')) {
                        hx = hx * 16 + strtol(&c, NULL, 16); // dig
                    } else break;
                    y++;
                }
                if (y == ysav) {
                    escape = true; // nothex
                    break;
                }
                if (mode == 's') {
                    mem[st] = hx;
                    st++;
                    continue;
                }
                if (mode == 'x') { // notstor
                    st = hx; // setadr
                    xam = hx;
                }
                bool addr = true;
                while (true) {
                    if (mode == 'x') { // hack
                        if (addr) // nxtprnt
                            printf("\n%04x:", xam);
                        printf(" %02x", mem[xam]); // prdata
                    }
                    mode = 'x'; // xamnext
                    if (xam >= hx) break;
                    xam++;
                    addr = xam % 8 == 0; // mod8chk
                }
            }
        }
    }
    return 0;
}
